### ISSUE TITLE
Connection string needs to support config StorageFeatures and DefaultPairs

### ISSUE THREAD

**Issue by Xuanwo** [beyondstorage/go-storage#680]

**Comment by xxchan** [comment]
a possible format:
- `storage_features=loose_pair, virtual_dir`
- `default_storage_pairs.write=[key1=value&key2=value]`

Implementation problems
- The syntax: 
  - We used `&` to split pairs, and disallow `&` in key and value. If `&` is used, we should rewrite parser. 
  - The key were pair names. If we want to use the dot syntax, we should rewrite parser.
- `*Features` & `Default*Pairs` are types defined in services. We cannot use them as the parsed value type. So `go-service-*` should register either a parse function or `reflect.Type`(?)
- `default_storage_pairs` needs the ability to parse pairs. So this cannot be parsed within the service, we should either:
  - register `reflect.Type`
  - register a parse function which accepts another parse function
  - export pair parsing API

**Comment by Xuanwo** [comment]
@JinnyYi, can you help continue the design?

**Comment by JinnyYi** [comment]
> @JinnyYi, can you help continue the design?

OK, I'll give feedback in time if there's any progress or questions.

**Comment by JinnyYi** [comment]
> * default_storage_pairs.write=[key1=value&key2=value]
> * A possible other compatible format is like `default_storage_pairs={write:[key1=value,key2=value],read:[]}` (a little bit ugly)

I'm not sure about the format for `default*pairs`. 
How about `default_storage_pairs.write=[key1=value, key2=value]&default_storage_pairs.write_append=[key1=value, key2=value]`? 
 It's similar to the first one. But they don't fit "the key is the pair name defined in toml and the format SHOULD be exactly the same" described in GSP-90?

**Comment by Xuanwo** [comment]
> But they don't fit "the key is the pair name defined in toml and the format SHOULD be exactly the same" described in GSP-90?

It's OK to loosen the requirements.

> How about `default_storage_pairs.write=[key1=value, key2=value]&default_storage_pairs.write_append=[key1=value, key2=value]`?

I prefer `default_storage_pairs={write:[storage_class=STANDARD_IA,key2=value],read:[]}`

---

`default_storage_pairs` exposes too many internal details to users.

Maybe we don't need to support let users set up default pairs via connection strings. Instead, we can add new pairs like:

- default_storage_class => convert to write/write_append/wirte_pair storage class pair.
- default_sse_key => convert to read/write/write_append sse related keys.

After these changes, we will only expose basic types in the connection string and don't need to support so complex config format.

**Comment by JinnyYi** [comment]
> `default_storage_pairs` exposes too many internal details to users.

Yes, I agree.

> Instead, we can add new pairs like:
> 
> * `default_storage_class` => convert to write/write_append/wirte_pair storage class pair.
> * `default_sse_key` => convert to read/write/write_append sse related keys.

For these pairs, maybe the effective values should be maintained by services? And we need to handle the differences between services.
I wonder whether it's friendly for users. Users may not sure which operations will be affected after setting the default pairs.

**Comment by Xuanwo** [comment]
> For these pairs, maybe the effective values should be maintained by services?

Nice question.

Maybe we should not handle them in go-storage globally, instead, we need to add them to every service. And the behavior should be defined via different services' RFCs.

>  I wonder whether it's friendly for users. Users may not sure which operations will be affected after setting the default pairs.

To me, I expect all operations should be affected. (Maybe we can add a detailed behavior description for the pair.)

**Comment by JinnyYi** [comment]
How about support config StorageFeatures first if needed? 

> * default_xxx (provided by different services)

Maybe we need a new topic to further discuss on the practicality of the demand(or feature) and specify which pairs need to and can be set as default ones?

**Comment by Xuanwo** [comment]
> How about support config StorageFeatures first if needed?

How about handle `StorageFeatures` in the same way? Like, add `enable_virtual_dir` pair?

> Maybe we need a new topic to further discuss on the practicality of the demand(or feature) and specify which pairs need to and can be set as default ones?

I think it's OK to discuss here, because they are focusing the same issue.

Once we reach the agreement, we can start an RFC process.

**Comment by JinnyYi** [comment]
> > How about support config StorageFeatures first if needed?
> 
> How about handle `StorageFeatures` in the same way? Like, add `enable_virtual_dir` pair?

Sorry for my mistake. I intended to confirm whether we should support `DefaultStoragePairs` first.

I have handled `StorageFeatures` based on `storage_features=loose_pair, virtual_dir` but not submitted. I can modify it according to `add `enable_virtual_dir` pair`.

**Comment by JinnyYi** [comment]
> * default_xxx (provided by different services)

Given that the format of the default pairs should have the prefix `default_` and follow with the pair name defined in toml. Maybe we can either:

- For all the operations, if they have the same required or optional pairs with the default ones, then we can set the pairs with the default value to the operations.

- We can refer to the relatively fixed fields in the common request headers to select the pairs that can be set as default ones, and we may need to sort the default pairs by operation or feature(Perhaps it's difficult to synchronize). Like:
  - `default_content_type`, `default_storage_class` for the requests(operations) that need to submit data, will affect write operations.
  - `default_server_side_encryption` for uploaded data(data encryption), will affect write operations.

**Comment by Xuanwo** [comment]
>     * For all the operations, if they have the same required or optional pairs with the default ones, then we can set the pairs with the default value to the operations.

I prefer this idea. How about in this way:

Add new field `defaultable` for operations, in which service can specify pairs can could be set as default:

```toml
[namespace.storage.op.write]
optional = ["content_md5", "content_type", "io_callback", "storage_class", "excepted_bucket_owner"]
default = ["excepted_bucket_owner"]

[namespace.storage.op.delete]
optional = ["excepted_bucket_owner", "multipart_id", "object_mode"]
default = ["excepted_bucket_owner"]
```

And we will generate `default_excepted_bucket_owner` pair for them.

- Operation has a different set of default pairs
- Pairs that have the same name will share the same default value.

More, we can generate the default logic in `parsePairServiceNew` and `parsePairStorageNew`, so that the service doesn't need to take any actions.

**Comment by JinnyYi** [comment]
I support your approach to specify pairs that can be set as default.

**Comment by Xuanwo** [comment]
> I support your approach to specify pairs that can be set as default.

Can you be responsible for drafting RFC and implementing this feature in its entirety?

**Comment by JinnyYi** [comment]
> Can you be responsible for drafting RFC and implementing this feature in its entirety?

Yes, of course.

**Comment by Xuanwo** [comment]
Fixed in #704
